File: openmate_agent/tool_monitor.py
from __future__ import annotations

import json
import math
from collections import defaultdict
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Literal
from uuid import uuid4

from pydantic import BaseModel, Field, model_validator

from openmate_shared.runtime_paths import resolve_workspace_root


ToolMonitorPhase = Literal["before", "after"]
ToolMonitorSource = Literal["model", "cli", "http", "unknown"]


def _utc_now() -> datetime:
    return datetime.now(UTC)


class ToolMonitorEvent(BaseModel):
    event_id: str = Field(min_length=1)
    phase: ToolMonitorPhase
    ts: datetime
    node_id: str = ""
    tool_name: str = ""
    source: ToolMonitorSource = "unknown"
    is_safe: bool = False
    is_read_only: bool = False
    request_id: str | None = None
    success: bool | None = None
    error_code: str | None = None
    duration_ms: int | None = None

    @model_validator(mode="after")
    def _validate_after_fields(self) -> "ToolMonitorEvent":
        if self.phase == "after" and self.success is None:
            raise ValueError("after event requires success")
        return self


class ToolMonitorStore:
    def __init__(self, file_path: str | Path) -> None:
        self._file_path = Path(file_path).resolve()

    @property
    def file_path(self) -> Path:
        return self._file_path

    def append(self, event: ToolMonitorEvent) -> None:
        self._file_path.parent.mkdir(parents=True, exist_ok=True)
        with self._file_path.open("a", encoding="utf-8", newline="\n") as handle:
            handle.write(event.model_dump_json(exclude_none=True))
            handle.write("\n")

    def list_all(self) -> list[ToolMonitorEvent]:
        if not self._file_path.exists():
            return []
        items: list[ToolMonitorEvent] = []
        with self._file_path.open("r", encoding="utf-8") as handle:
            for raw in handle:
                line = raw.strip()
                if not line:
                    continue
                try:
                    items.append(ToolMonitorEvent.model_validate_json(line))
                except Exception:
                    continue
        return items


class ToolMonitorService:
    def __init__(self, workspace_root: str | Path, store: ToolMonitorStore | None = None) -> None:
        root = resolve_workspace_root(workspace_root)
        default_path = root / ".openmate" / "runtime" / "tool_monitor.jsonl"
        self._store = store or ToolMonitorStore(default_path)

    @property
    def store(self) -> ToolMonitorStore:
        return self._store

    def record_before(
        self,
        *,
        node_id: str,
        tool_name: str,
        source: ToolMonitorSource,
        is_safe: bool,
        is_read_only: bool,
        request_id: str | None = None,
        ts: datetime | None = None,
    ) -> None:
        event = ToolMonitorEvent(
            event_id=str(uuid4()),
            phase="before",
            ts=ts or _utc_now(),
            node_id=node_id,
            tool_name=tool_name,
            source=source,
            is_safe=is_safe,
            is_read_only=is_read_only,
            request_id=request_id,
        )
        self._store.append(event)

    def record_after(
        self,
        *,
        node_id: str,
        tool_name: str,
        source: ToolMonitorSource,
        is_safe: bool,
        is_read_only: bool,
        success: bool,
        error_code: str | None,
        duration_ms: int,
        request_id: str | None = None,
        ts: datetime | None = None,
    ) -> None:
        event = ToolMonitorEvent(
            event_id=str(uuid4()),
            phase="after",
            ts=ts or _utc_now(),
            node_id=node_id,
            tool_name=tool_name,
            source=source,
            is_safe=is_safe,
            is_read_only=is_read_only,
            request_id=request_id,
            success=success,
            error_code=error_code,
            duration_ms=duration_ms,
        )
        self._store.append(event)

    def list_events(
        self,
        *,
        tool_name: str | None = None,
        node_id: str | None = None,
        source: ToolMonitorSource | None = None,
        success: bool | None = None,
        limit: int | None = None,
        window_minutes: int | None = None,
    ) -> list[ToolMonitorEvent]:
        events = self._store.list_all()
        if tool_name:
            events = [item for item in events if item.tool_name == tool_name]
        if node_id:
            events = [item for item in events if item.node_id == node_id]
        if source:
            events = [item for item in events if item.source == source]
        if success is not None:
            events = [item for item in events if item.phase == "after" and item.success == success]
        if window_minutes is not None and window_minutes > 0:
            cutoff = _utc_now() - timedelta(minutes=window_minutes)
            events = [item for item in events if item.ts >= cutoff]
        events.sort(key=lambda item: item.ts, reverse=True)
        if limit is not None and limit >= 0:
            return events[:limit]
        return events

    def summarize(
        self,
        *,
        tool_name: str | None = None,
        node_id: str | None = None,
        source: ToolMonitorSource | None = None,
        success: bool | None = None,
        limit: int | None = None,
        window_minutes: int | None = None,
    ) -> list[dict[str, float | int | str]]:
        events = self.list_events(
            tool_name=tool_name,
            node_id=node_id,
            source=source,
            success=success,
            limit=None,
            window_minutes=window_minutes,
        )
        after_events = [item for item in events if item.phase == "after"]
        grouped: dict[str, list[ToolMonitorEvent]] = defaultdict(list)
        for item in after_events:
            grouped[item.tool_name].append(item)

        summary: list[dict[str, float | int | str]] = []
        for name, items in grouped.items():
            count = len(items)
            success_count = sum(1 for item in items if item.success)
            durations = [item.duration_ms for item in items if item.duration_ms is not None]
            avg_duration = float(sum(durations) / len(durations)) if durations else 0.0
            p95_duration = float(self._percentile95(durations)) if durations else 0.0
            summary.append(
                {
                    "tool_name": name,
                    "count": count,
                    "success_rate": float(success_count / count) if count else 0.0,
                    "avg_duration_ms": avg_duration,
                    "p95_duration_ms": p95_duration,
                }
            )
        summary.sort(key=lambda item: (-int(item["count"]), str(item["tool_name"])))
        if limit is not None and limit >= 0:
            return summary[:limit]
        return summary

    @staticmethod
    def _percentile95(values: list[int]) -> int:
        if not values:
            return 0
        sorted_values = sorted(values)
        rank = int(math.ceil(0.95 * len(sorted_values))) - 1
        rank = max(0, min(rank, len(sorted_values) - 1))
        return sorted_values[rank]


def dump_events_json(events: list[ToolMonitorEvent]) -> list[dict[str, object]]:
    rows: list[dict[str, object]] = []
    for item in events:
        payload = item.model_dump(mode="json", exclude_none=True)
        if isinstance(payload.get("ts"), str):
            payload["ts"] = payload["ts"].replace("+00:00", "Z")
        rows.append(payload)
    return rows

File: internal/vos/httpapi/tool_monitor.go
package httpapi

import (
	"bufio"
	"encoding/json"
	"math"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"vos/internal/vos/domain"
)

type toolMonitorEvent struct {
	EventID    string  `json:"event_id"`
	Phase      string  `json:"phase"`
	TS         string  `json:"ts"`
	NodeID     string  `json:"node_id,omitempty"`
	ToolName   string  `json:"tool_name,omitempty"`
	Source     string  `json:"source,omitempty"`
	IsSafe     bool    `json:"is_safe"`
	IsReadOnly bool    `json:"is_read_only"`
	RequestID  *string `json:"request_id,omitempty"`
	Success    *bool   `json:"success,omitempty"`
	ErrorCode  *string `json:"error_code,omitempty"`
	DurationMS *int    `json:"duration_ms,omitempty"`
}

type toolMonitorQuery struct {
	ToolName      string
	NodeID        string
	Source        string
	Success       *bool
	Limit         int
	WindowMinutes int
}

type toolMonitorSummaryItem struct {
	ToolName      string  `json:"tool_name"`
	Count         int     `json:"count"`
	SuccessRate   float64 `json:"success_rate"`
	AvgDurationMS float64 `json:"avg_duration_ms"`
	P95DurationMS float64 `json:"p95_duration_ms"`
}

func (server *Server) handleV1ToolMonitorRoutes(writer http.ResponseWriter, request *http.Request) {
	path := strings.TrimPrefix(request.URL.Path, v1Prefix+"/tools/monitor")
	switch path {
	case "/events":
		server.handleV1ToolMonitorEvents(writer, request)
	case "/summary":
		server.handleV1ToolMonitorSummary(writer, request)
	default:
		server.writeV1Error(writer, http.StatusNotFound, "not found")
	}
}

func (server *Server) handleV1ToolMonitorEvents(writer http.ResponseWriter, request *http.Request) {
	if request.Method != http.MethodGet {
		server.writeV1MethodNotAllowed(writer, request.Method, http.MethodGet)
		return
	}
	query, err := parseToolMonitorQuery(request)
	if err != nil {
		server.writeV1ServiceError(writer, err)
		return
	}
	events, err := server.loadToolMonitorEvents()
	if err != nil {
		server.writeV1ServiceError(writer, err)
		return
	}
	filtered := filterToolMonitorEvents(events, query, false)
	server.writeV1Success(writer, filtered)
}

func (server *Server) handleV1ToolMonitorSummary(writer http.ResponseWriter, request *http.Request) {
	if request.Method != http.MethodGet {
		server.writeV1MethodNotAllowed(writer, request.Method, http.MethodGet)
		return
	}
	query, err := parseToolMonitorQuery(request)
	if err != nil {
		server.writeV1ServiceError(writer, err)
		return
	}
	events, err := server.loadToolMonitorEvents()
	if err != nil {
		server.writeV1ServiceError(writer, err)
		return
	}
	filtered := filterToolMonitorEvents(events, query, true)
	summary := summarizeToolMonitorEvents(filtered, query.Limit)
	server.writeV1Success(writer, summary)
}

func parseToolMonitorQuery(request *http.Request) (toolMonitorQuery, error) {
	values := request.URL.Query()
	query := toolMonitorQuery{
		ToolName: strings.TrimSpace(values.Get("tool_name")),
		NodeID:   strings.TrimSpace(values.Get("node_id")),
		Source:   strings.TrimSpace(values.Get("source")),
		Limit:    100,
	}

	if query.Source != "" {
		switch query.Source {
		case "model", "cli", "http", "unknown":
		default:
			return toolMonitorQuery{}, domain.ValidationError{Message: "source must be one of model|cli|http|unknown"}
		}
	}

	if raw := strings.TrimSpace(values.Get("success")); raw != "" {
		parsed, err := strconv.ParseBool(raw)
		if err != nil {
			return toolMonitorQuery{}, domain.ValidationError{Message: "success must be true or false"}
		}
		query.Success = &parsed
	}
	if raw := strings.TrimSpace(values.Get("limit")); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil || parsed <= 0 {
			return toolMonitorQuery{}, domain.ValidationError{Message: "limit must be a positive integer"}
		}
		query.Limit = parsed
	}
	if raw := strings.TrimSpace(values.Get("window_minutes")); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil || parsed <= 0 {
			return toolMonitorQuery{}, domain.ValidationError{Message: "window_minutes must be a positive integer"}
		}
		query.WindowMinutes = parsed
	}
	return query, nil
}

func (server *Server) loadToolMonitorEvents() ([]toolMonitorEvent, error) {
	path := filepath.Join(server.workspace, filepath.FromSlash(".openmate/runtime/tool_monitor.jsonl"))
	file, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return []toolMonitorEvent{}, nil
		}
		return nil, err
	}
	defer file.Close()

	events := []toolMonitorEvent{}
	scanner := bufio.NewScanner(file)
	buffer := make([]byte, 0, 64*1024)
	scanner.Buffer(buffer, 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var event toolMonitorEvent
		if err := json.Unmarshal([]byte(line), &event); err != nil {
			continue
		}
		events = append(events, event)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return events, nil
}

func filterToolMonitorEvents(events []toolMonitorEvent, query toolMonitorQuery, afterOnly bool) []toolMonitorEvent {
	cutoff := time.Time{}
	if query.WindowMinutes > 0 {
		cutoff = time.Now().UTC().Add(-time.Duration(query.WindowMinutes) * time.Minute)
	}

	filtered := make([]toolMonitorEvent, 0, len(events))
	for _, event := range events {
		if afterOnly && event.Phase != "after" {
			continue
		}
		if query.ToolName != "" && event.ToolName != query.ToolName {
			continue
		}
		if query.NodeID != "" && event.NodeID != query.NodeID {
			continue
		}
		if query.Source != "" && event.Source != query.Source {
			continue
		}
		if query.Success != nil {
			if event.Phase != "after" || event.Success == nil || *event.Success != *query.Success {
				continue
			}
		}
		if !cutoff.IsZero() {
			parsed, err := time.Parse(time.RFC3339Nano, strings.TrimSpace(event.TS))
			if err != nil || parsed.Before(cutoff) {
				continue
			}
		}
		filtered = append(filtered, event)
	}

	sort.SliceStable(filtered, func(i, j int) bool {
		left, leftErr := time.Parse(time.RFC3339Nano, strings.TrimSpace(filtered[i].TS))
		right, rightErr := time.Parse(time.RFC3339Nano, strings.TrimSpace(filtered[j].TS))
		if leftErr == nil && rightErr == nil {
			return left.After(right)
		}
		return filtered[i].TS > filtered[j].TS
	})

	if query.Limit > 0 && len(filtered) > query.Limit {
		return filtered[:query.Limit]
	}
	return filtered
}

func summarizeToolMonitorEvents(events []toolMonitorEvent, limit int) []toolMonitorSummaryItem {
	grouped := map[string][]toolMonitorEvent{}
	for _, event := range events {
		if event.Phase != "after" {
			continue
		}
		grouped[event.ToolName] = append(grouped[event.ToolName], event)
	}

	summary := make([]toolMonitorSummaryItem, 0, len(grouped))
	for toolName, items := range grouped {
		count := len(items)
		if count == 0 {
			continue
		}
		successCount := 0
		durations := make([]int, 0, count)
		for _, item := range items {
			if item.Success != nil && *item.Success {
				successCount++
			}
			if item.DurationMS != nil {
				durations = append(durations, *item.DurationMS)
			}
		}
		avgDuration := 0.0
		p95Duration := 0.0
		if len(durations) > 0 {
			total := 0
			for _, duration := range durations {
				total += duration
			}
			avgDuration = float64(total) / float64(len(durations))
			p95Duration = float64(percentile95(durations))
		}
		summary = append(summary, toolMonitorSummaryItem{
			ToolName:      toolName,
			Count:         count,
			SuccessRate:   float64(successCount) / float64(count),
			AvgDurationMS: avgDuration,
			P95DurationMS: p95Duration,
		})
	}

	sort.Slice(summary, func(i, j int) bool {
		if summary[i].Count == summary[j].Count {
			return summary[i].ToolName < summary[j].ToolName
		}
		return summary[i].Count > summary[j].Count
	})
	if limit > 0 && len(summary) > limit {
		return summary[:limit]
	}
	return summary
}

func percentile95(values []int) int {
	if len(values) == 0 {
		return 0
	}
	sortedValues := append([]int(nil), values...)
	sort.Ints(sortedValues)
	rank := int(math.Ceil(0.95*float64(len(sortedValues)))) - 1
	if rank < 0 {
		rank = 0
	}
	if rank >= len(sortedValues) {
		rank = len(sortedValues) - 1
	}
	return sortedValues[rank]
}
